### DISCUSSION TITLE
Fast link force algorithm

### DISCUSSION THREAD

**Discussion by loadingcn** [cosmograph-org/cosmos#91] [Ideas]
Hi, I have implement a fast version of link force calculation here:
https://github.com/loadingcn/cosmos/commit/f41787ea3f5883273f196eb783b81b3997362ade

Basically, the vertex shader treat each link as a computing task, it add the result to the velocity of source node.
But the code did not run well, in some case, it simply run in a mass after some iterations.
Is there anything I missed in the algorithm?

**Comment by Stukova** [comment]
@loadingcn Thank you for your fast version of link force calculation. Your version is great, we also tried to implement a similar one, but ran into a problem as I see your version also has one.

Your version sums up small `dx` and `dy` vectors into a node texture. The total vector can grow unpredictably large, especially if there are many links connected to the node. This version is indeed optimal and more logical for a parallel approach in shaders, but it has a potential issue when dealing with link forces.

In the context of link forces, it is important to apply to node the force from each link sequentially, so that the next force applied depends on the previous force applied to the node. This is similar to a classical approach. Unfortunately, I couldn't think of a way to implement this in an optimal parallel approach 🤔.

**Reply by loadingcn** [reply]
Thanks, your explanation is quite clear 👍!
It seems I had to figure out another way to make it run fast.
